internal/locate/cse/cse.go

```diff
@@ -0,0 +1,428 @@
+
+	mu struct {
+		sync.RWMutex
+		stores map[uint64]*store
+	}
+
+	httpClient *http.Client
+
+	gp *util.Spool
+}
+
+func NewCSEClient(origin pd.Client) (c *Client, err error) {
+	c = &Client{
+		Client: origin,
+
+		done: make(chan struct{}, 1),
+
+		httpClient: &http.Client{
+			Transport: &http.Transport{},
+		},
+
+		gp: util.NewSpool(32, time.Second*10),
+	}
+	c.mu.stores = make(map[uint64]*store)
+	err = c.refreshStores()
+	if err != nil {
+		return nil, err
+	}
+	go c.refreshStoresLoop()
+	return c, nil
+}
+
+func (c *Client) probeStoreStatus(name string, addr string, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("http://%s", addr), nil)
+	if err != nil {
+		return err
+	}
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		log.Warn("failed to check store", zap.String("name", name), zap.Error(err))
+		return err
+	}
+	defer resp.Body.Close()
+	log.Warn("mark store reachable", zap.String("name", name), zap.String("store", addr))
+	return nil
+}
+
+func (c *Client) updateStores(stores []*metapb.Store) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for _, s := range stores {
+		if origin, ok := c.mu.stores[s.Id]; !ok {
+			s := &store{
+				Store: s,
+			}
+			settings := settings{
+				Name:          fmt.Sprintf("store-%d", s.GetId()),
+				Interval:      5 * time.Second,
+				Timeout:       1 * time.Second,
+				ProbeInterval: 1 * time.Second,
+				ReadyToTrip:   ifMostFailures,
+				Probe: func(name string) error {
+					addr := s.GetStatusAddress()
+					log.Warn("store is marked as unavailable, probing",
+						zap.String("name", name),
+						zap.String("store", addr))
+					return c.probeStoreStatus(name, addr, 1*time.Second)
+				},
+			}
+			s.breaker = newAsyncBreaker(settings)
+			c.mu.stores[s.GetId()] = s
+		} else {
+			if s.GetState() == metapb.StoreState_Tombstone {
+				origin.breaker.Close()
+				delete(c.mu.stores, s.GetId())
+				continue
+			}
+			origin.Store = s
+		}
+	}
+}
+
+func (c *Client) getAllStores() []*metapb.Store {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	var stores []*metapb.Store
+	for _, s := range c.mu.stores {
+		stores = append(stores, s.Store)
+	}
+	return stores
+}
+
+func (c *Client) refreshStores() error {
+	stores, err := c.Client.GetAllStores(context.Background(), pd.WithExcludeTombstone())
+	if err != nil {
+		return err
+	}
+	c.updateStores(stores)
+	return nil
+}
+
+func (c *Client) refreshStoresLoop() {
+	ticker := time.NewTicker(StoresRefreshInterval)
+	for {
+		select {
+		case <-c.done:
+			return
+		case <-ticker.C:
+			err := c.refreshStores()
+			if err != nil {
+				log.Error("refresh stores failed", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (c *Client) Close() {
+	// Cancel all the ongoing requests.
+	c.gp.Close()
+	// Close the idle connections.
+	c.httpClient.CloseIdleConnections()
+	// Close the stores refresh goroutine.
+	c.done <- struct{}{}
```

Comment:
how about `close(c.done)` ? That will not block the current goroutinue.

Reply:
This will not block the current goroutine since the channel is init with: `done: make(chan struct{}, 1)`